commit 87dad1a13ceff2ffc2d6e268d8da24d2ddce9340
Author: Raphael <68374617+raphaelscholle@users.noreply.github.com>
Date:   Fri Mar 29 05:41:38 2024 +0100

    remove incompatible wifi

diff --git a/kernel/nvidia/drivers/net/wireless/Kconfig b/kernel/nvidia/drivers/net/wireless/Kconfig
index 1a0f2f620..d90bb8d42 100644
--- a/kernel/nvidia/drivers/net/wireless/Kconfig
+++ b/kernel/nvidia/drivers/net/wireless/Kconfig
@@ -1,8 +1,4 @@
 source "drivers/net/wireless/bcmdhd/Kconfig"
 source "drivers/net/wireless/bcmdhd_pcie/Kconfig"
 source "drivers/net/wireless/realtek/rtl8812au/Kconfig"
-source "drivers/net/wireless/realtek/rtl8814au/Kconfig"
-source "drivers/net/wireless/realtek/rtl8821au/Kconfig"
-source "drivers/net/wireless/realtek/rtl8821cu/Kconfig"
-source "drivers/net/wireless/realtek/rtl8822bu/Kconfig"
-source "drivers/net/wireless/realtek/rtl8822ce/Kconfig"
+source "drivers/net/wireless/realtek/rtl88x2bu/Kconfig"
diff --git a/kernel/nvidia/drivers/net/wireless/Makefile b/kernel/nvidia/drivers/net/wireless/Makefile
index e19a0949b..8359a2fd5 100644
--- a/kernel/nvidia/drivers/net/wireless/Makefile
+++ b/kernel/nvidia/drivers/net/wireless/Makefile
@@ -1,8 +1,4 @@
 obj-$(CONFIG_BCMDHD_SDIO) += bcmdhd/
 obj-$(CONFIG_BCMDHD_PCIE) += bcmdhd_pcie/
 obj-$(CONFIG_RTL8812AU) += realtek/rtl8812au/
-obj-$(CONFIG_RTL8814AU) += realtek/rtl8814au/
-obj-$(CONFIG_RTL8821AU) += realtek/rtl8821au/
-obj-$(CONFIG_RTL8821CU) += realtek/rtl8821cu/
-obj-$(CONFIG_RTL8822BU) += realtek/rtl8822bu/
-obj-$(CONFIG_RTL8822CE) += realtek/rtl8822ce/
+obj-$(CONFIG_RTL8822BU) += realtek/rtl88x2bu/
diff --git a/kernel/kernel-5.10/arch/arm64/configs/defconfig b/kernel/kernel-5.10/arch/arm64/configs/defconfig
index 56311d9fd..f22ccd015 100644
--- a/kernel/kernel-5.10/arch/arm64/configs/defconfig
+++ b/kernel/kernel-5.10/arch/arm64/configs/defconfig
@@ -616,8 +616,10 @@ CONFIG_RTL8188EE=m
 CONFIG_RTL8192EE=m
 CONFIG_RTL8821AE=m
 CONFIG_RTL8192CU=m
-CONFIG_RTL8XXXU=m
-CONFIG_RTL8XXXU_UNTESTED=y
+CONFIG_RTL8XXXU=n
+CONFIG_RTL8XXXU_UNTESTED=n
+CONFIG_RTL8812AU=m
+CONFIG_RTL8822BU=n
 CONFIG_RSI_91X=m
 # CONFIG_RSI_COEX is not set
 CONFIG_CW1200=m
diff --git a/kernel/kernel-5.10/drivers/net/wireless/realtek/rtl8812au/Makefile b/kernel/kernel-5.10/drivers/net/wireless/realtek/rtl8812au/Makefile
index e5eeaf08e..7584859d7 100644
--- a/kernel/kernel-5.10/drivers/net/wireless/realtek/rtl8812au/Makefile
+++ b/kernel/kernel-5.10/drivers/net/wireless/realtek/rtl8812au/Makefile
@@ -305,7 +305,7 @@ _BTC_FILES += hal/btc/halbtc8192e1ant.o \
 endif
 
 ifeq ($(CONFIG_PLATFORM_ARM64_RPI), y)
-include $(TopDIR)/drivers/net/wireless/rtl8812au/hal/phydm/phydm.mk
+include $(TopDIR)/../kernel/kernel-5.10/drivers/net/wireless/realtek/rtl8812au/hal/phydm/phydm.mk
 else
 include $(TopDIR)/hal/phydm/phydm.mk
 endif
